### PR TITLE
[WIP] Add WebWorker API

### PR DESCRIPTION
**Pull request by bartlomieju**
Ref #1222 

ATM I am getting this error during build:
```
Undefined symbols for architecture x86_64:
  "v8::Shell::WorkerTerminate(v8::FunctionCallbackInfo<v8::Value> const&)", referenced from:
      deno::CreateGlobalTemplate(v8::Isolate*) in libdeno.a(binding.o)
  "v8::Shell::WorkerGetMessage(v8::FunctionCallbackInfo<v8::Value> const&)", referenced from:
      deno::CreateGlobalTemplate(v8::Isolate*) in libdeno.a(binding.o)
  "v8::Shell::WorkerPostMessage(v8::FunctionCallbackInfo<v8::Value> const&)", referenced from:
      deno::CreateGlobalTemplate(v8::Isolate*) in libdeno.a(binding.o)
  "v8::Shell::WorkerNew(v8::FunctionCallbackInfo<v8::Value> const&)", referenced from:
      deno::CreateGlobalTemplate(v8::Isolate*) in libdeno.a(binding.o)
```

### CODE REVIEW

**Comment by ry** [comment]
Interesting!

v8::Shell functions have to be copied over. Start a new file called libdeno/worker.cc and add them. You can grep for the locations in third_party/v8

**Comment by bartlomieju** [comment]
I started with `WorkerNew`. It required copying a lot of code from `v8/src/d8.cc` and I can't get rid of those two errors:
```
In file included from ../../libdeno/binding.cc:13:
../../libdeno/worker.cc:108:42: error: incomplete type 'v8::internal::String' named in nested name specifier
  if (i::FLAG_use_external_strings && i::String::IsAscii(chars, size)) {
                                      ~~~^~~~~~~~
../../third_party/v8/src/globals.h:528:7: note: forward declaration of 'v8::internal::String'
class String;
      ^
In file included from ../../libdeno/binding.cc:13:
../../libdeno/worker.cc:187:14: error: no matching member function for call to 'push_back'
    workers_.push_back(worker);
    ~~~~~~~~~^~~~~~~~~
../../third_party/llvm-build/Release+Asserts/include/c++/v1/vector:707:36: note: candidate function not viable: no known conversion from 'v8::Worker *' to 'const std::__1::__vector_base<deno::Worker *, std::__1::allocator<deno::Worker *> >::value_type' (aka 'deno::Worker *const') for 1st argument
    _LIBCPP_INLINE_VISIBILITY void push_back(const_reference __x);
                                   ^
../../third_party/llvm-build/Release+Asserts/include/c++/v1/vector:710:36: note: candidate function not viable: no known conversion from 'v8::Worker *' to 'std::__1::vector<deno::Worker *, std::__1::allocator<deno::Worker *> >::value_type' (aka 'deno::Worker *') for 1st argument
    _LIBCPP_INLINE_VISIBILITY void push_back(value_type&& __x);
```


Is that the right direction @ry?

**Comment by ry** [comment]
Ya - you will have to make some heavy modifications to get it to work. I will give some more detailed comments later

**Comment by bartlomieju** [comment]
Awesome!

**Comment by kevinkassimo** [comment]
Just curious (sorry for not really following the discussion about workers recently in Node), are we going to have separate event loops for each worker thread, or are we going to share a single event loop across all worker threads?

Also I believe we also need to implement MessagePorts together with Workers

**Comment by kitsonk** [comment]
It _should_ match [the specification](https://www.w3.org/TR/workers/#the-event-loop) IMO, which is a seperate event loop.

File: libdeno/api.cc [hunk 1]
```diff
@@ -30,12 +30,13 @@ Deno* deno_new(deno_buf snapshot, deno_buf shared, deno_recv_cb cb) {
   v8::Locker locker(isolate);
   v8::Isolate::Scope isolate_scope(isolate);
   {
+    v8::Local<v8::ObjectTemplate> global_template =
+        deno::CreateGlobalTemplate(isolate);
```
Comment (ry) [inline libdeno/api.cc]: This works. I would consider using our existing `deno::InitializeContext` for setting up the global variables... However using a ObjectTemplate is probably a better option - and how D8 is doing it - so don't worry about this nit for now.

File: libdeno/worker.cc [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+  fclose(file);
+  *size_out = static_cast<int>(size);
+  return chars;
+}
+
+
+// Reads a file into a v8 string.
+v8::Local<v8::String> ReadFile(v8::Isolate* isolate, const char* name) {
+  int size = 0;
+  char* chars = ReadChars(name, &size);
```
Comment (ry) [inline libdeno/worker.cc]: file_util.cc can be used here to read a file.

File: libdeno/worker.cc [hunk 3]
```diff
@@ -0,0 +1,255 @@
+#include <errno.h>
```
Comment (ry) [inline libdeno/worker.cc]: Please maintain the copyright header from V8 in this file.

File: libdeno/worker.cc [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+      : data_(std::move(data)), length_(length) {}
+  const char* data() const override { return data_.get(); }
+  size_t length() const override { return length_; }
+
+ private:
+  std::unique_ptr<const char[]> data_;
+  size_t length_;
+};
+
+static FILE* FOpen(const char* path, const char* mode) {
```
Comment (ry) [inline libdeno/worker.cc]: This can be removed if you use file_util.h

File: libdeno/worker.cc [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  if (fstat(fileno(file), &file_stat) != 0) return nullptr;
+  bool is_regular_file = ((file_stat.st_mode & S_IFREG) != 0);
+  if (is_regular_file) return file;
+  fclose(file);
+  return nullptr;
+#endif
+}
+
+
+static char* ReadChars(const char* name, int* size_out) {
```
Comment (ry) [inline libdeno/worker.cc]: This can be removed if you use file_util.h

**Comment by bartlomieju** [comment]
@ry I've started with your feedback. 

I put some placeholder methods (like `Worker::ExecuteInThread`) to get compiler happy, but that resulted in:
```

[----------] 17 tests from LibDenoTest
[ RUN      ] LibDenoTest.InitializesCorrectly

#
# Fatal error in v8::HandleScope::CreateHandle()
# Cannot create a handle without a HandleScope
#
```

Digging further!

**Comment by ry** [comment]
@bartlomieju In V8's binding API, basically every function needs to declare a HandleScope for storing local variables. I suggest reading https://v8.dev/docs/embed
For debugging such errors, if a line number isn't printed, you should use a debugger like lldb or gdb. I often debug deno like this:
```
lldb -- target/debug/deno tests/worker.js
> run
> bt
> up
> up
> l
```

**Comment by ry** [comment]
@kevinkassimo Sorry didn't see your message until now. I'd like the different workers to share the same Tokio runtime.

**Comment by bartlomieju** [comment]
Thanks for tips @ry great read. 

There is some progress now, actual worker instantiation is called (run `deno tests/test_worker.ts`) but some error when reading worker file emerges. I'll get to that tomorrow and keep going.

**Comment by ry** [comment]
I've started an alternative implementation in rust here #1331

**Comment by bartlomieju** [comment]
Closing in favor of #1331

**Pull request closed without merging**